fieldservice_contract/models/contract_line.py

```diff
@@ -0,0 +1,251 @@
+        :rtype dict
+        """
+        result = {}
+        for line in self:
+            # create fsm_recurring
+            values = line._field_create_fsm_recurring_prepare_values()
+            fsm_recurring = self.env["fsm.recurring"].sudo().create(values)
+            # fsm_recurring.action_start()  # do not start yet
+            line.write({"fsm_recurring_id": fsm_recurring.id})
+            result[line.id] = fsm_recurring
+        return result
+
+    def _field_find_fsm_order(self):
+        """Find the fsm_order generated by the lines. If no fsm_order
+        linked, it will be created automatically.
+        :return a mapping with the line id and its linked fsm_order
+        :rtype dict
+        """
+        # one search for all lines
+        fsm_orders = self.env["fsm.order"].search(
+            [("contract_line_id", "in", self.ids)]
+        )
+        fsm_order_cl_mapping = {
+            fsm_order.contract_line_id.id: fsm_order for fsm_order in fsm_orders
+        }
+        result = {}
+        for line in self:
+            # If the contract was confirmed, cancelled,
+            # set to draft then confirmed,
+            # avoid creating a new fsm_order.
+            fsm_order = fsm_order_cl_mapping.get(line.id)
+            # If not found, create one fsm_order for the line
+            if not fsm_order:
+                fsm_order = line._field_create_fsm_order()[line.id]
+            result[line.id] = fsm_order
+        return result
+
+    def _field_find_fsm_recurring(self):
+        """Find the fsm_recurring generated by the lines. If no
+        fsm_recurring linked, it will be created automatically.
+        :return a mapping with the line id and its linked
+        fsm_recurring
+        :rtype dict
+        """
+        # one search for all lines
+        fsm_recurrings = self.env["fsm.recurring"].search(
+            [("contract_line_id", "in", self.ids)]
+        )
+        fsm_recurring_cl_mapping = {
+            fsm_recurring.contract_line_id.id: fsm_recurring
+            for fsm_recurring in fsm_recurrings
+        }
+        result = {}
+        for line in self:
+            # If the contract was confirmed, cancelled,
+            # set to draft then confirmed,
+            # avoid creating a new fsm_recurring.
+            fsm_recurring = fsm_recurring_cl_mapping.get(line.id)
+            # If not found, create one fsm_recurring for the line
+            if not fsm_recurring:
+                fsm_recurring = line._field_create_fsm_recurring()[line.id]
+            result[line.id] = fsm_recurring
+        return result
+
+    def _field_service_generation(self):
+        """For service lines, create the field service order. If it already
+        exists, it simply links the existing one to the line.
+        """
+        for line in self:
+            # create order
+            if line.product_id.field_service_tracking == "order":
```

Comment:
Field_service_tracking has not "order" value https://github.com/OCA/field-service/blob/14.0/fieldservice_sale/models/product_template.py#L17.  We have "sale" and "line" value which mean that we can create an unique FSM order by sale order if Field_service_tracking = "sale" or one SFM Order by sale line if Field_service_tracking = "line".
It seam that for recurring order we can only create one FSM recurring order by sale line.